src/settings/global_user.rs

```diff
@@ -17,7 +26,90 @@ pub enum GlobalUser {
 
 impl GlobalUser {
     pub fn new() -> Result<Self, failure::Error> {
-        get_global_config()
+        let environment = Environment::with_whitelist(ENV_VAR_WHITELIST.to_vec());
+
+        let config_path = default_config_file().expect("could not find global config directory");
+
+        GlobalUser::build(environment, config_path)
+    }
+
+    fn build<T: 'static + QueryEnvironment>(
+        environment: T,
+        config_path: PathBuf,
+    ) -> Result<Self, failure::Error>
+    where
+        T: config::Source + Send + Sync,
+    {
+        if let Some(user) = Self::from_env(environment) {
+            user
+        } else {
+            Self::from_file(config_path)
+        }
+    }
+
+    // TODO: This fn should return None if the environment has none of the
+    // relevant variables
+    fn from_env<T: 'static + QueryEnvironment>(
+        environment: T,
+    ) -> Option<Result<Self, failure::Error>>
+    where
+        T: config::Source + Send + Sync,
+    {
+        if environment.empty().unwrap() {
```

Comment:
why do we need an `unwrap` here? is it a `Result`? In what scenarios would `environment.empty()` return an error, and can we use `?` to bubble up the error instead of unwrapping it?

Reply:
it's not a result, it's an option.

Reply:
env::var() has a few different errors. perhaps we should match on some of those in `collect`? because the most likely error is that the env var simply doesn't exist, and that's not an error for us.

Reply:
an `.unwrap_or(true)`. incoming.